x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/RepositoryAnalyzeAction.java

```diff
@@ -394,7 +396,9 @@ public AsyncAction(
             this.discoveryNodes = discoveryNodes;
             this.currentTimeMillisSupplier = currentTimeMillisSupplier;
             this.timeoutTimeMillis = currentTimeMillisSupplier.getAsLong() + request.getTimeout().millis();
-            this.listener = listener;
+
+            this.cancellationListener = new SubscribableListener<>();
+            this.listener = ActionListener.runBefore(listener, () -> cancellationListener.onResponse(null));
```

Comment:
I am trying to think through the difference scenarios how `cancellationListener` can already be completed before we invoke `onResponse(null)` here. I think they are all OK. I am writing it down to be explicit and maybe you can double check it as well.
1. The request fails due to timeout, i.e. `cancellationListener` is already timed out and this runs right before we are going to call `listener.onFailure`. This is fine because `SubscribableListener` accepts only the first completion and silently ignores all future results.
2. We are about to call `listener.onResponse` for success while `cancellationListener` times out concurrently. The timeout will set the failure and try to cancel the tasks. This is fine because we don't check the failure object anymore and cancelling completed or non-existing tasks seem to be a noop.
3. Timeout can be called even after the `listener` is completed. This is fine since the timeout will be after `cancellationListener.onResponse(null)` and completing a `SubscribableListener` more than once is ignored.

Reply:
:+1: sounds about right, yes.